BUG: inconsistent DataFrame.agg behavoir when passing as kwargs numeric_only=True

thlautenschlaeger opened the pull request:
- [x] closes #49352 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Commit:
add func list handling for numeric_only

thlautenschlaeger commented:
@rhshadrach can you please have a look at the PR since v1.5.2 will be released.

jreback commented:
this doesn't have any tests

thlautenschlaeger commented:
I wasn't sure if they are covered by default. Will add them. Thanks for your response.

thlautenschlaeger commented:
To check in advance, is the appropriate location to add the test in the script below?:

`pandas/tests/apply/test_frame_apply.py`

rhshadrach commented on pandas/core/apply.py:
```diff
@@ -723,13 +723,18 @@ def apply(self) -> DataFrame | Series:
         # TODO: Avoid having to change state
         self.obj = self.obj if self.axis == 0 else self.obj.T
         self.axis = 0
+        self._filter_numeric_only()
```
This applies the numeric_only in all cases rather that just where it's necessary. Can you move this to agg_list_like and agg_dict_like.

thlautenschlaeger replied:
Good point. I will move it.

rhshadrach commented:
> To check in advance, is the appropriate location to add the test in the script below?:
> 
> `pandas/tests/apply/test_frame_apply.py`

Yes - that's correct.

rhshadrach reviewed (commented):
Thanks for the PR!

For the overall the behavior here, I didn't realize commenting in the issue that this approach would be something of a hidden feature. That is, numeric_only isn't in the arguments of the function, but we're essentially implementing it as if it were. 

I think this would be (at least more) okay if we were only exercising `numeric_only` when it came to string arguments. E.g. ['sum', lambda x, numeric_only: x.sum(), 'mean']` would do the computation on the numeric portion of the DataFrame for 'sum' and 'mean', but would attempt to pass every Series that makes up the frame to the lambda. This is admittedly not great - I think the only proper resolution is to use DataFrame methods in agg for list/dict like arguments, but that can't happen until 2.0.

cc @mroeschke for any thoughts.

mroeschke commented:
If it were better documented that passing the string alias of the DataFrame methods accepts the associated keyword arguments then I supposed it makes sense to support this, especially if the user passes multiple strings aliases into agg.

Commit:
added tests

Commit:
added issue id to tests

rhshadrach commented:
Thanks @mroeschke - @thlautenschlaeger can you add a description of whats happening here to the notes section of agg and apply.

Commit:
fixed test

Commit:
fixed typing issue

Commit:
added notes for documentation

Commit:
fixed docstring

thlautenschlaeger commented:
> Thanks @mroeschke - @thlautenschlaeger can you add a description of whats happening here to the notes section of agg and apply.

Added the description to that section.  Also added an example to the doc for agg. Hope everything is alright now.

mroeschke commented on pandas/core/frame.py:
```diff
@@ -9142,6 +9142,23 @@ def _gotitem(
+    ...                   columns=['A', 'B', 'C'])
+
+    Works equivalently as above. Add argument `numeric_only=True` to avoid
+    exceptions or warnings.
```
Nit: The warning will go away soon (in 2.0) so I am not sure it's worth including here

mroeschke commented on doc/source/whatsnew/v1.5.2.rst:
```diff
@@ -27,7 +27,7 @@ Fixed regressions
 ~~~~~~~~~
 - Bug in the Copy-on-Write implementation losing track of views in certain chained indexing cases (:issue:`48996`)
--
+- Bug in inconsistent DataFrame.agg behavior when passing as kwargs ``numeric_only=True`` (:issue:`49352`
```
```suggestion
- Bug in inconsistent DataFrame.agg behavior when passing as kwargs ``numeric_only=True`` with :class:`DataFrame` methods or their associated string alias as functions (:issue:`49352`)
```

mroeschke reviewed (commented):
@rhshadrach I'm assuming the other kwargs like `skipna` and `min_periods` have the same issue here in `agg`, but those don't need to be addressed in 1.5.x since those don't affect the `FutureWarning`?

rhshadrach reviewed (changes requested):
This is looking good - a few requests below. Also, when mentioning the `numeric_only` in the docs, can you state it applies only to string arguments. That it works for e.g. `DataFrame.mean` is part of the standard behavior of `agg` (to pass any keywords through to the provided function), so I don't think that needs to be particularly called out.

rhshadrach commented:
> @rhshadrach I'm assuming the other kwargs like `skipna` and `min_periods` have the same issue here in `agg`, but those don't need to be addressed in 1.5.x since those don't affect the `FutureWarning`?

Correct. While I do think things like skipna should work, they are not part of the regression due to the numeric_only changes.

rhshadrach commented on pandas/core/apply.py:
```diff
@@ -368,8 +377,11 @@ def agg_dict_like(self) -> DataFrame | Series:
+
         obj = self.obj
         arg = cast(AggFuncTypeDict, self.f)
+        arg = {k: arg[k] for k in arg.keys() if k not in filtered_col}
```
I specified modifying agg_dict_like before, but I'm now thinking I'm that we shouldn't be.  E.g. `b` a non-numeric-column:

    df.agg({"a": "sum", "b": "mean"}, numeric_only=True)

I don't think there is any expectation that you can pass `numeric_only` here. Apologies for sending you down the wrong path - unless I'm missing something.

thlautenschlaeger commented:
Thanks for your reviews @mroeschke @rhshadrach 
I will go through your comments and integrate the change requests.

Commit:
applied patch

Commit:
applied patch

Commit:
added change to whats new

rhshadrach commented on pandas/core/apply.py:
```diff
@@ -284,6 +283,13 @@ def transform_str_or_callable(self, func) -> DataFrame | Series:
             return func(obj, *args, **kwargs)
 
+    def _filter_numeric_only(self) -> list[Any]:
+        if "numeric_only" in self.kwargs and bool(self.kwargs["numeric_only"]) is True:
```
Instead of 

    and bool(self.kwargs["numeric_only"]) is True 

why not just 

    and self.kwargs["numeric_only"]

rhshadrach commented on pandas/core/apply.py:
```diff
@@ -294,7 +300,9 @@ def agg_list_like(self) -> DataFrame | Series:
         from pandas.core.reshape.concat import concat
 
-        obj = self.obj
+        filtered_cols = self._filter_numeric_only()
+        n = len(filtered_cols)
+        obj = self.obj if n == 0 else self.obj[filtered_cols].astype("O")
```
Why the conversion to object type here? Can `_filter_numeric_only` just return obj?

Since _filter_numeric_only doesn't always filter, can you rename to _maybe_filter_numeric_only

rhshadrach commented on pandas/core/frame.py:
```diff
@@ -9213,6 +9213,23 @@ def _gotitem(
+    Works equivalently as above. Add argument `numeric_only=True` to
+    aggregate only numeric columns.
+
+    >>> df.agg({'B': ['mean', 'std'], 'C': ['sum', 'mean']},
+    ...                    numeric_only=True)
```
Can you use a list example here

rhshadrach commented on pandas/core/frame.py:
```diff
@@ -9377,6 +9394,9 @@ def apply(
         for more details.
+        Use the keyword argument `numeric_only=True` to apply functions
+        only to numeric columns and to skip the non-numeric columns.
+        e.g. the column contains a string.
```
nit: either use `columns, e.g. the column` or capitalize the 'e' in 'e.g.'

rhshadrach commented on pandas/core/shared_docs.py:
```diff
@@ -46,6 +46,10 @@
+
+Use the keyword argument `numeric_only=True` to apply functions
+only to numeric columns and to skip the non-numeric columns.
+e.g. the column contains a string.
```
Same as above

rhshadrach reviewed (changes requested):
I think we need to only filter the DataFrame due to `numeric_only` on string arguments in the list-like. Otherwise, we will be creating an inconsistency between list and callable arguments. For example, with `f` a callable user-defined function

    df.agg(f, numeric_only=True)

will pass `numeric_only` to the function `f` whereas

    df.agg([f], numeric_only=True)

will filter the DataFrame prior to calling `f`.

Commit:
resolved threads

thlautenschlaeger committed:
Merge branch 'main' into fix_numeric_list_func

thlautenschlaeger committed:
fix doc

rhshadrach commented on pandas/core/apply.py:
```diff
@@ -284,6 +284,14 @@ def transform_str_or_callable(self, func) -> DataFrame | Series:
             return func(obj, *args, **kwargs)
 
+    def _filter_numeric_only(self) -> list[Any]:
+        if "numeric_only" in self.kwargs and self.kwargs["numeric_only"] is True:
```
I don't believe there is any need for the `is True` here. This would give surprising results if the user passes e.g. `np.bool_(True)`.

Also, can you use `kwargs.get("numeric_only", False)`

rhshadrach commented on pandas/core/frame.py:
```diff
@@ -9213,6 +9213,21 @@ def _gotitem(
+    ...                    ['c', 3, 6]],
+    ...                   columns=['A', 'B', 'C'])
+
+    Works equivalently as above. Add argument `numeric_only=True` to
```
What does "Works equivalently as above." mean here?

rhshadrach commented on pandas/core/apply.py:
```diff
@@ -294,7 +299,7 @@ def agg_list_like(self) -> DataFrame | Series:
         from pandas.core.reshape.concat import concat
 
-        obj = self.obj
+        obj = self._maybe_filter_numeric_only()
```
This will cause different behaviors for list and non-list arguments that contain a callable. E.g.

```
df.agg(lambda x: x.sum(), numeric_only=True))
df.agg([lambda x: x.sum()], numeric_only=True)
```

Currently in this PR, I believe the first line will pass `numeric_only` to the supplied UDF (and this fails as the UDF doesn't take a numeric_only argument). The second line will subset the DataFrame to only numeric columns and succeed. These should have the same behavior with regards to how the numeric_only argument is treated.

I suggest only using the DataFrame filtered to numeric_only when a given argument in the list is a string; otherwise to use the full DataFrame and pass the numeric_only kwarg to the supplied UDF.

rhshadrach commented on pandas/tests/apply/test_frame_apply.py:
```diff
@@ -858,6 +858,22 @@ def test_with_dictlike_columns_with_infer():
     tm.assert_frame_equal(result, expected)
 
 
+def test_with_dictlike_functions():
```
We're no longer changing dictlike behavior, I think this can be removed. It should fail, but that is another (unrelated) issue.

rhshadrach commented:
@thlautenschlaeger - friendly ping. Are you interested in continuing?

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
